core/src/main/scala/sttp/tapir/metrics/Metric.scala

```diff
@@ -0,0 +1,15 @@
+package sttp.tapir.metrics
+
+import sttp.tapir.Endpoint
+import sttp.tapir.model.{ServerRequest, ServerResponse}
+
+import scala.concurrent.duration.Deadline
+
+case class Metric[M](
+    metric: M,
+    onRequest: Option[(Endpoint[_, _, _, _], ServerRequest, M) => Unit] = None,
```

Comment:
idea: maybe a metric could be initialised per-request (in the request interceptor) using `onRequest: (ServerRequest, M) => P`, where `P` is the payload type. This payload would then be passed into `onEndpointRequest: (Endpoint, ServerRequest, M, P) => F[Unit]` & similarly to `onEndpointResponse`.

This would allow us to generalise over the quite specific `Deadline` parameter.

Reply:
And another question: shouldn't the metric function return `F[Unit]`, not `Unit`? Capturing a metric could be a side-effecting operation (it probably is ;) )

Reply:
Or maybe simpler, we could go back to the idea of "initialising" a metric on request start, which would produce a "proper" metric (a constant instance in most cases, but for capturing timings this would include the initial timestamp). So here we'd only have a function `onRequest: (ServerRequest, M) => EndpointMetric`